### ISSUE TITLE
GitLab Support

### ISSUE THREAD

**Issue by fkorotkov** [cirruslabs/cirrus-ci-docs#10]

**Comment by yzgyyang** [comment]
👍

**Comment by RDIL** [comment]
Any updates on this?  I now have a self-hosted instance that I would love to use Cirrus CI with.

**Comment by fkorotkov** [comment]
No updates so far. Since people prefer to host GitLab in most of the cases. This feature will follow after self-hosted Cirrus.

**Comment by sio** [comment]
I've built an integrated GitLab/Cirrus CI pipeline using the GraphQL API:

- Blog post: https://potyarkin.ml/posts/2020/cirrus-ci-integration-for-gitlab-projects/
- Underlying tool: https://github.com/sio/cirrus-run

Native integration would've been better, but this approach is almost as convenient. 

Let me know what do you think 😊

**Comment by fkorotkov** [comment]
@sio wow! That's an elegant workaround! Thanks for sharing! 

Do you have a Twitter account? Would love to share the blog post and mention you from [@cirrus_labs Twitter account](https://twitter.com/cirrus_labs).

**Comment by sio** [comment]
@fkorotkov, no, I'm not on Twitter. Linking the blog post or GitHub project is perfectly ok.

**Comment by sio** [comment]
I was very excited to learn that [Libvirt](https://gitlab.com/libvirt/libvirt) team uses Cirrus CI (via GitLab via cirrus-run) to execute their FreeBSD and macOS builds [for a couple of months already](https://www.mail-archive.com/libvir-list@redhat.com/msg201902.html)! They have even ditched Travis in favor of Cirrus CI!

**Comment by AlexWayfer** [comment]
> This feature will follow after self-hosted Cirrus.

Is here an issue for it?

**Comment by berrange** [comment]
> I was very excited to learn that [Libvirt](https://gitlab.com/libvirt/libvirt) team uses Cirrus CI (via GitLab via cirrus-run) to execute their FreeBSD and macOS builds [for a couple of months already](https://www.mail-archive.com/libvir-list@redhat.com/msg201902.html)!

We (libvirt) have used this very successfully for a good time now, and the [QEMU](https://gitlab.com/qemu-project/qemu) project has adopted it too.

There is one key downside to the cirrus-run approach for integrating GitLab and Cirrus CI. The cirrus-run client executes in the GitLab CI job, and all the time that it waits around for completion of the Cirrus CI job, is counted against the quota for GitLab CI, even though cirrus-run is essentially idle during this waiting time. When GitLab switches to enforcing quotas on CI usage in the near future, this is going to be a significant downside and will likely force libvirt & QEMU to stop using cirrus-run. 

We really appreciate that Cirrus CI provides CI runners for free for macOS and FreeBSD. If Cirrus CI were to gain native integration with GitLab for reporting CI results, it would be amazing.

**Comment by berrange** [comment]
FYI, GitLab has now dramatically reduced the number of CI minutes available to GitLab CI pipelines, by several orders of magnitude. As an unfortunate result, the cirrus-run approach to triggering Cirrus CI jobs from GitLab is now much less sustainable/viable, because it is burning GitLab CI minutes while waiting for the Cirrus jobs to complete.  So native Cirrus integration with GitLab is more desirable now than in the past.

**Comment by hasufell** [comment]
3 years later and GithHub becoming more aggressive towards users with self-hosted runners: so what's the status?

This would essentially make cirrus CI a tool to liberate users from the platform lock-in of github/gitlab etc. We could freely migrate without having to rewrite our CI all the time. Surprised Cirrus doesn't see this as a massive opportunity.
